### PR TITLE
Merge quantile sketches as they arrive

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
Currently, the quantile sketch merge logic in the query frontend waits for all shards to complete and merges them then. That puts a lot of pressure on the memory. Luckily the merges can happen out of order as they come in.

```
› benchstat before2.log after.log
goos: linux
goarch: amd64
pkg: github.com/grafana/loki/pkg/querier/queryrange
cpu: AMD Ryzen 7 3700X 8-Core Processor             
                                 │ before2.log │             after.log              │
                                 │   sec/op    │   sec/op     vs base               │
Accumulator/streams-16             46.51µ ± 2%   45.74µ ± 2%       ~ (p=0.280 n=10)
Accumulator/quantile_sketches-16   6.806m ± 2%   7.012m ± 0%  +3.02% (p=0.000 n=10)
geomean                            562.7µ        566.3µ       +0.65%

                                 │  before2.log  │              after.log               │
                                 │     B/op      │     B/op      vs base                │
Accumulator/streams-16              8.385Ki ± 0%   8.384Ki ± 0%        ~ (p=0.570 n=10)
Accumulator/quantile_sketches-16   68.482Ki ± 0%   4.465Ki ± 1%  -93.48% (p=0.000 n=10)
geomean                             23.96Ki        6.119Ki       -74.47%

                                 │ before2.log │              after.log              │
                                 │  allocs/op  │ allocs/op   vs base                 │
Accumulator/streams-16              79.00 ± 0%   79.00 ± 0%       ~ (p=1.000 n=10) ¹
Accumulator/quantile_sketches-16    69.00 ± 3%   67.00 ± 1%  -2.90% (p=0.010 n=10)
geomean                             73.83        72.75       -1.46%
¹ all samples are equal
```

I've simulated the later merge. 

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/querier/queryrange/downstreamer.go [hunk 1]
```diff
@@ -297,6 +295,36 @@ func (a *bufferedAccumulator) Result() []logqlmodel.Result {
...
+	}
+
+	var err error
+	a.matrix, err = a.matrix.Merge(data)
+	return err
+}
+
+func (a *quantileSketchAccumulator) Result() []logqlmodel.Result {
+	return []logqlmodel.Result{{Data: a.matrix}}
+}
```
Comment (jeschkies) [inline pkg/querier/queryrange/downstreamer.go:326]: Should this be the place where we merge results?

**Comment by jeschkies** [comment]
I still need to benchmark this.

**Comment by cstyan** [comment]
I've pushed two commits to your branch, one with benchmarks for single merge vs accumulator and one for an optimization to the map usage in the quantile matrix/vector merging.

First looking at just the single merge as we have it prior to the accumulator change, with and without the pooling:
```
cpu: AMD Ryzen 9 5950X 16-Core Processor
               │   no-pool   │                pool                │
               │   sec/op    │   sec/op     vs base               │
SingleMerge-32   5.032m ± 6%   3.846m ± 5%  -23.56% (p=0.002 n=6)

               │     no-pool     │                pool                 │
               │      B/op       │     B/op      vs base               │
SingleMerge-32   1043.875Ki ± 0%   3.390Ki ± 3%  -99.68% (p=0.002 n=6)

               │   no-pool   │               pool                │
               │  allocs/op  │ allocs/op   vs base               │
SingleMerge-32   1771.0 ± 0%   129.5 ± 0%  -92.69% (p=0.002 n=6)
```

Then the pooled single merge vs accumulator merge:
```
     │ AccumulateMerge │         SingleMerge          │
     │     sec/op      │   sec/op     vs base         │
*-32       3.932m ± 2%   3.846m ± 5%  ~ (p=0.093 n=6)

     │ AccumulateMerge │             SingleMerge              │
     │      B/op       │     B/op      vs base                │
*-32      1.051Ki ± 4%   3.390Ki ± 3%  +222.58% (p=0.002 n=6)

     │ AccumulateMerge │             SingleMerge             │
     │    allocs/op    │  allocs/op   vs base                │
*-32        30.00 ± 3%   129.50 ± 0%  +331.67% (p=0.002 n=6)
```

And the final result with my two commits vs the single merge usage:
```
     │ SingleMerge │          AccumulateMerge           │
     │   sec/op    │   sec/op     vs base               │
*-32   5.032m ± 6%   3.932m ± 2%  -21.86% (p=0.002 n=6)

     │   SingleMerge   │           AccumulateMerge           │
     │      B/op       │     B/op      vs base               │
*-32   1043.875Ki ± 0%   1.051Ki ± 4%  -99.90% (p=0.002 n=6)

     │ SingleMerge  │          AccumulateMerge          │
     │  allocs/op   │ allocs/op   vs base               │
*-32   1771.00 ± 0%   30.00 ± 3%  -98.31% (p=0.002 n=6)
```

File: pkg/querier/queryrange/downstreamer_test.go [hunk 2]
```diff
@@ -598,3 +600,70 @@ func TestDownstreamAccumulatorMultiMerge(t *testing.T) {
...
+		})
+	}
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		acc := newQuantileSketchAccumulator()
+		for _, r := range res {
+			require.NoError(b, acc.Accumulate(r, 0))
+		}
+	}
+}
```
Comment (jeschkies) [inline pkg/querier/queryrange/downstreamer_test.go:639]: Nice. I was about to push my change and had a conflict. Your code was almost exactly as mine :tada:

File: pkg/querier/queryrange/downstreamer.go [hunk 3]
```diff
@@ -248,7 +238,8 @@ func newDownstreamAccumulator(params logql.Params, nQueries int) *downstreamAccu
 }
 
 func (a *downstreamAccumulator) build(acc logqlmodel.Result) {
```
Comment (jeschkies) [inline pkg/querier/queryrange/downstreamer.go:240]: I'm not too fond of this design. We do know whether the results are samples, logs or quantile sketches. So ideally the logql engine would inject the accumulator on the execution plan. Moving this into the logql package would also make the `MockQuerier` simpler.

Comment (cstyan) [reply pkg/querier/queryrange/downstreamer.go:240]: this seems like a good change to me

**Review by cstyan — APPROVED** [review]

**Pull request merged**